# Lab | Inferential statistics - ANOVA

**Part 1**

Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

**Context**

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data anova_lab_data.xlsx file in the files_for_lab folder.

**Part 2**

    In this section, use Python to conduct ANOVA.
    What conclusions can you draw from the experiment and why?


## Part 1

### Steps to set up ANOVA    
    
    - null hypothesis: power = same avg etching rate
    - alternate hypothesis: power != same avg etching rate
    - confidence level: 0.95
    - significance level: 0.05 (one-way ANOVA)
    - degrees of freedom of model: 2 numerator (nr. of groups - 1 = DoF)
    - total degrees of freedom: 15 - 3 = 12 denominator (total observation - nr. of groups = total DoF)
    - critical value: 3.8853 (see f-table)
    

## Part 2

### Load libraries

In [5]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

### Load data

In [4]:
data = pd.read_excel(r'C:/Users/josefin/01_IRONHACK/Week8/Day8.3/lab-inferential-statistics-anova/files_for_lab/anova_lab_data.xlsx', sheet_name='data_collected')
data.columns = ['power', 'etching_rate']
data

power  etching_rate
0   160 W          5.43
1   180 W          6.24
2   200 W          8.79
3   160 W          5.71
4   180 W          6.71
5   200 W          9.20
6   160 W          6.22
7   180 W          5.98
8   200 W          7.90
9   160 W          6.01
10  180 W          5.66
11  200 W          8.15
12  160 W          5.59
13  180 W          6.60
14  200 W          7.55

### Compute global mean and means of each group

In [7]:
# Global mean:
gm = np.mean(data['etching_rate'])
print("Global mean",gm)

# Mean of every group:
group_averages = [(group,np.mean(data[data['power'] == group]['etching_rate'])) for group in list(data['power'].unique())] 
print("The group means are: ",group_averages)

Global mean 6.782666666666667
The group means are:  [('160 W', 5.792), ('180 W', 6.2379999999999995), ('200 W', 8.318)]


### Compute numerator, denominator & ratio

In [8]:
numerator = 0
for group in group_averages:
    numerator += len(data[data['power'] == group[0]]['etching_rate'].values) * ( group[1] - gm )**2
numerator = numerator / ( len(data['power'].unique()) - 1 ) 

print("\hat{S}_{t}^{2} is: ", numerator)

\hat{S}_{t}^{2} is:  9.088326666666667


In [11]:
denominator = 0
for group in group_averages: 
    for value in data[data['power'] == group[0]]['etching_rate'].values:
        denominator += (value - group[1])**2
denominator = denominator / ( len(data) - len(data['power'].unique()) )

print("\hat{S}_{e}^{2} is: ",denominator)

\hat{S}_{e}^{2} is:  0.24643666666666644


In [12]:
F = numerator / denominator

print("The F statistic is: ",F)

The F statistic is:  36.87895470100503


### Look up critical value

critical value for F = 3.8853

### Conclusion

As our F-statistic (36.88) > F-critical value (3.89), **we reject the H0**.

### ALTERNATIVE: automatic computation with stats module from scipy

In [15]:
import scipy.stats as stats

stats.f_oneway(data[data['power'] == '160 W']['etching_rate'],
               data[data['power'] == '180 W']['etching_rate'],
               data[data['power'] == '200 W']['etching_rate'])

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)